File: spigot-impl-hidden-details/src/main/java/us/minevict/mvutil/spigot/hidden/CloseDatabaseListener.java
package us.minevict.mvutil.spigot.hidden;

import co.aikar.idb.Database;
import java.util.HashMap;
import java.util.Map;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.server.PluginDisableEvent;
import org.bukkit.plugin.Plugin;
import org.jetbrains.annotations.NotNull;

public final class CloseDatabaseListener implements Listener {
  private static final Object LOCK = new Object();
  private static CloseDatabaseListener INSTANCE = null;

  @NotNull
  private final Map<Plugin, Database> databases = new HashMap<>();

  private CloseDatabaseListener() {
  }

  @NotNull
  public static CloseDatabaseListener getInstance() {
    if (INSTANCE == null) {
      synchronized (LOCK) {
        if (INSTANCE == null) {
          return INSTANCE = new CloseDatabaseListener();
        }
      }
    }

    return INSTANCE;
  }

  @EventHandler(priority = EventPriority.MONITOR)
  public void onPluginDisable(PluginDisableEvent event) {
    var database = databases.remove(event.getPlugin());
    if (database != null) {
      event.getPlugin().getLogger().info("MV-Util closing database for plugin.");
      database.close();
    }
  }

  @NotNull
  public Map<Plugin, Database> getDatabases() {
    return databases;
  }
}

File: spigot/src/main/kotlin/us/minevict/mvutil/spigot/internal/CloseDatabaseListener.kt
/**
 * MV-Util
 * Copyright (C) 2020 Mariell Hoversholm, Nahuel Dolores
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package us.minevict.mvutil.spigot.internal

import co.aikar.idb.Database
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.server.PluginDisableEvent
import org.bukkit.plugin.Plugin

internal object CloseDatabaseListener : Listener {
    val LOCK = Any()
    val databases = mutableMapOf<Plugin, Database>()

    @EventHandler(priority = EventPriority.MONITOR)
    fun disable(event: PluginDisableEvent) {
        val db = databases[event.plugin] ?: return
        event.plugin.logger.info("MV-Util closing database for plugin.")
        db.close()
    }
}